### ISSUE TITLE
Find Async, listener never called

### ISSUE THREAD

**Issue by Dean79000** [realm/realm-java#4396]
# Goal

Check if object not exist in db ASYNC

(work with synchronous .findFirst() return me `null` - the expected result)
(work also if i call `.load()`)

Listener / query seams be not registered for updates / results.
 
# Expected Results

ChangeListener called with null or `loaded`=true + `valid`=false

# Actual Results

Changelistener never called, and `loaded` always = false

# Steps & Code to Reproduce

```java      
final Ressource res2 = Realm.getDefaultInstance().where(Ressource.class).equalTo("id", "not_exist").findFirstAsync();
res2.addChangeListener(new RealmChangeListener<RealmModel>() {
    @Override
    public void onChange(RealmModel element) {
        Log.e("this", "never called !");
    }
});
```

See .zip  to reproduce problem : [project demo.zip](https://github.com/realm/realm-java/files/879269/project.demo.zip)

# Version of Realm and tooling

- Realm version(s): realm-gradle-plugin:3.0.0
- Realm sync feature enabled: no
- Android Studio version:  Android Studio 2.3
- Build #AI-162.3764568, built on February 24, 2017
- JRE: 1.8.0_112-release-b06 amd64
- JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
- Which Android version and device:  API 25 - Official Emulator (Haxm 6.0.5)

**Comment by Zhuinden** [comment]
You need to keep `Ressource` as a field variable of the class you're in, or add it to a list and then remove it when the change listener is called, stuff like that.

**Comment by Dean79000** [comment]
This is the same, i have simplify the exemple.

**Comment by Zhuinden** [comment]
But

```
final Ressource res2 = Realm.getDefaultInstance().where(Ressource.class).equalTo("id", "not_exist").findFirstAsync();
res2.addChangeListener(
```

This won't work, because `res2` is not a field variable of the class.

It's a local variable in the method.

**Comment by istx25** [comment]
Hi @Dean79000. Wanted to chime in here. Please try @Zhuinden's suggestion and if you're still experiencing issues, please follow-up here and we'll get one of the Realm Java engineers to review what you've shared and help resolve the issue. Thanks.

**Comment by beeender** [comment]
@Dean79000 as @Zhuinden said, when `res2` gets GCed, all the listeners on the `res2` will be GCed as well.

**Comment by Dean79000** [comment]
I have updated the code, `res2` with class variable, even with static, still experiencing issue.
[project demo v2.zip](https://github.com/realm/realm-java/files/881234/project.demo.v2.zip)

**Comment by beeender** [comment]
@Dean79000 I am very sorry there was a wrong behaviour for `findFirstAsync()` before `v3.1.0` which is fixed in #4352 . Before `v3.1.0` the query will keep running until it can find an object match the condition.
The fix has been merged, so you can try with our snapshot release. (I tried with your project, it works fine :) )

See https://github.com/realm/realm-java#using-snapshots about how to use snapshot release.

BTW, what @Zhuinden said is still correct in the new version, you still have to keep a ref to `res2` to ensure the listener is valid.

**Comment by Zhuinden** [comment]
Oh yeah, that was `findFirstAsync` in the question. My eyes skipped it after seeing the local ref :sad:
